src/main/java/cn/nukkit/entity/item/EntityVehicle.java

```diff
@@ -1,46 +1,124 @@
-    public static final int DATA_HURT_DIRECTION = 18;
+public abstract class EntityVehicle extends EntityInteractable implements EntityRideable {
+
+    // Ahm this wrong...
+    // Maybe someone can get the (REAL) one?
     public static final int DATA_DAMAGE_TAKEN = 19;
 
     public EntityVehicle(FullChunk chunk, CompoundTag nbt) {
         super(chunk, nbt);
     }
 
-    public int getHurtTime() {
-        return this.getDataPropertyInt(EntityVehicle.DATA_HURT_TIME);
+    public int getRollingAmplitude() {
+        return this.getDataPropertyInt(DATA_HURT_TIME);
     }
 
-    public void setHurtTime(int time) {
-        this.setDataProperty(new IntEntityData(EntityVehicle.DATA_HURT_TIME, time));
+    public void setRollingAmplitude(int time) {
+        this.setDataProperty(new IntEntityData(DATA_HURT_TIME, time));
     }
 
-    public int getHurtDirection() {
-        return this.getDataPropertyInt(EntityVehicle.DATA_HURT_DIRECTION);
+    public int getRollingDirection() {
+        return this.getDataPropertyInt(DATA_HURT_DIRECTION);
     }
 
-    public void setHurtDirection(int direction) {
-        this.setDataProperty(new IntEntityData(EntityVehicle.DATA_HURT_DIRECTION, direction));
+    public void setRollingDirection(int direction) {
+        this.setDataProperty(new IntEntityData(DATA_HURT_DIRECTION, direction));
     }
 
-    public float getDamageTaken() {
+    public float getDamage() {
         return this.getDataPropertyFloat(EntityVehicle.DATA_DAMAGE_TAKEN);
     }
 
-    public void setDamageTaken(float damage) {
+    public void setDamage(float damage) {
         this.setDataProperty(new FloatEntityData(EntityVehicle.DATA_DAMAGE_TAKEN, damage));
     }
+
+    @Override
+    public String getInteractButton() {
+        return "mount";
+    }
+
+    @Override
+    public boolean canDoInteraction() {
+        return linkedEntity != null;
+    }
+
+    @Override
+    public boolean onUpdate(int currentTick) {
+        if (justCreated) {
+            VehicleCreateEvent event = new VehicleCreateEvent(this);
+            getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                kill();
+                return false;
+            }
+            justCreated = false;
+        }
+
+        // The rolling amplitude
+        if (getRollingAmplitude() > 0) {
+            setRollingAmplitude(getRollingAmplitude() - 1);
+        }
+        // The damage token
+        if (getDamage() > 0) {
+            setDamage(getDamage() - 1);
+        }
+        // A killer task
+        if (y < -16) {
+            kill();
+        }
+        // Movement code
+        updateMovement();
+        return true;
+    }
+
+    public boolean performHurtAnimation(float damage) {
+        // Vehicle does not respond hurt animation on packets
+        // It only respond on vehicle data flags. Such as these
+        setRollingAmplitude(10);
+        setRollingDirection(renderSideBySideAnimation());
```

Comment:
rolling direction is only 1 or -1 anything else and rolling amplitude seems to be good, the only thing that MCPE sends extra is health data